libkbfs/folder_block_ops.go

```diff
@@ -1256,45 +1310,135 @@ func (fbo *folderBlockOps) Write(
-func (fbo *folderBlockOps) truncateLocked(
+// truncateExtendLocked is called by truncateLocked to extend a file and
+// creates a hole.
+func (fbo *folderBlockOps) truncateExtendLocked(
 	ctx context.Context, lState *lockState, md *RootMetadata,
-	file path, size uint64) (*WriteRange, []BlockPointer, error) {
-	fbo.blockLock.AssertLocked(lState)
+	file path, size uint64) (WriteRange, []BlockPointer, error) {
 
-	// check writer status explicitly
-	username, uid, err := fbo.config.KBPKI().GetCurrentUserInfo(ctx)
+	if size > fbo.config.MaxFileBytes() {
+		return WriteRange{}, nil, FileTooBigError{file, int64(size), fbo.config.MaxFileBytes()}
+	}
+
+	fblock, uid, err := fbo.writeGetFileLocked(ctx, lState, md, file)
 	if err != nil {
-		return nil, nil, err
+		return WriteRange{}, nil, err
 	}
-	if !md.GetTlfHandle().IsWriter(uid) {
-		return nil, nil, NewWriteAccessError(md.GetTlfHandle(), username)
+
+	var dirtyPtrs []BlockPointer
+
+	fbo.log.CDebugf(ctx, "truncateExtendLocked: extending fblock %#v", fblock)
+	if !fblock.IsInd {
+		fbo.log.CDebugf(ctx, "truncateExtendLocked: making block indirect %v", file.tailPointer())
+		old := fblock
+		fblock, err = fbo.createIndirectBlockLocked(md, file, uid,
+			DefaultNewBlockDataVersion(fbo.config, true))
+		if err != nil {
+			return WriteRange{}, nil, err
+		}
+		fblock.IPtrs[0].Holes = true
+		err = fbo.cacheBlockIfNotYetDirtyLocked(lState, fblock.IPtrs[0].BlockPointer, file.Branch, old)
+		if err != nil {
+			return WriteRange{}, nil, err
+		}
+		dirtyPtrs = append(dirtyPtrs, fblock.IPtrs[0].BlockPointer)
+		fbo.log.CDebugf(ctx, "truncateExtendLocked: new zero data block %v", fblock.IPtrs[0].BlockPointer)
 	}
 
-	fblock, err := fbo.getFileLocked(ctx, lState, md, file, blockWrite)
+	// TODO: support multiple levels of indirection.  Right now the
+	// code only does one but it should be straightforward to
+	// generalize, just annoying
+
+	err = fbo.newRightBlockLocked(ctx, lState, file.tailPointer(),
+		file.Branch, fblock, int64(size), md)
 	if err != nil {
-		return nil, nil, err
+		return WriteRange{}, nil, err
+	}
+	dirtyPtrs = append(dirtyPtrs, fblock.IPtrs[len(fblock.IPtrs)-1].BlockPointer)
+	fbo.log.CDebugf(ctx, "truncateExtendLocked: new right data block %v", fblock.IPtrs[len(fblock.IPtrs)-1].BlockPointer)
+
+	de, err := fbo.getDirtyEntryLocked(ctx, lState, md, file)
+	if err != nil {
+		return WriteRange{}, nil, err
+	}
+
+	si := fbo.getOrCreateSyncInfoLocked(lState, de)
+
+	de.EncodedSize = 0
+	// update the file info
+	de.Size = size
+	fbo.deCache[file.tailPointer().ref()] = de
+
+	// Always make the top block dirty, so we will sync its
+	// indirect blocks.  This has the added benefit of ensuring
+	// that any write to a file while it's being sync'd will be
+	// deferred, even if it's to a block that's not currently
+	// being sync'd, since this top-most block will always be in
+	// the fileBlockStates map.
+	for i := range fblock.IPtrs {
```

Comment:
The comment right above this refers to the `cacheBlockIfNotYetDirtyLocked` call below.  Should probably rearrange it so the comment is immediately above that call.

Also, this loop could use a separate comment.  I assume we're setting `Holes` to `true` on all of the pointers as a shortcut -- we only really need to set it on the one representing the hole.  The shortcut is fine -- I just think it should be commented.